ipld/kamt/src/iter.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+use std::borrow::Borrow;
+use std::iter::FusedIterator;
+
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::de::DeserializeOwned;
+use fvm_ipld_encoding::CborStore;
+
+use crate::hash_bits::HashBits;
+use crate::node::{match_extension, ExtensionMatch, Node};
+use crate::pointer::Pointer;
+use crate::{AsHashedKey, Config, Error, KeyValuePair};
+#[doc(hidden)]
+pub struct IterImpl<'a, BS, V, K, H, const N: usize = 32> {
+    store: &'a BS,
+    stack: Vec<std::slice::Iter<'a, Pointer<K, V, H, N>>>,
+    current: std::slice::Iter<'a, KeyValuePair<K, V>>,
+}
+
+pub type Iter<'a, BS, V, K, H, const N: usize = 32> = IterImpl<'a, BS, V, K, H, N>;
+impl<'a, K, V, BS, H, const N: usize> IterImpl<'a, BS, V, K, H, N>
+where
+    K: DeserializeOwned,
+    V: DeserializeOwned,
+    BS: Blockstore,
+{
+    pub(crate) fn new(store: &'a BS, root: &'a Node<K, V, H, N>) -> Self {
+        Self {
+            store,
+            stack: vec![root.pointers.iter()],
+            current: [].iter(),
+        }
+    }
+
+    pub(crate) fn new_from<Q: Sized>(
+        store: &'a BS,
+        root: &'a Node<K, V, H, N>,
+        key: &Q,
+        conf: &'a Config,
+    ) -> Result<Self, Error>
+    where
+        K: Borrow<Q>,
+        Q: PartialEq,
+        H: AsHashedKey<Q, N>,
+    {
+        println!("IterImpl::new_from");
+        let hashed_key = H::as_hashed_key(key);
+        let mut hash = HashBits::new(&hashed_key);
+        let mut node = root;
+        let mut ext;
```

Comment:
This should be inside the loop (doesn't cost anything extra and keeps us from keeping it around between iterations).

Reply:
And, inside the loop, it doesn't need to be `mut`.